### PR TITLE
Bugfix for generate page from db table and add analytics event after the successful CRUD page creation

### LINKED ISSUE #6086: [Bug] Find query fails for generate page from DB table flow for MongoDB datasource
## Description

Create a CRUD page from generate page from DB flow. On landing to the created page find query was failing. 

### Steps to reproduce the behaviour:

1. Generate CRUD page from MongoDB datasource
2. Find query fails when user lands on created page with error in action configuration setting

### PR DESCRIPTION
**Pull request by abhvsn**
## Description

> Find query was failing for MongoDB datasource because of the _where_ operator, which is linked with Select Widget. Allocation of labels and values were incorrect earlier for MongoDB datasource. Also added a analytics datapoint after successful page creation.   

Fixes #6086

## Type of change

- Bug fix (non-breaking change which fixes an issue)
- New feature (non-breaking change which adds functionality)

## How Has This Been Tested?

> Tested manually on local instance

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java [hunk 1]
```diff
@@ -81,12 +87,12 @@
 
     private final String LIST_QUERY = "ListFiles";
 
+    private final String S3_PLUGIN_PACKAGE = "amazons3-plugin";
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:90]: I suggest to use `private static final String` here and for other final constants.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java [hunk 2]
```diff
@@ -96,7 +102,7 @@
     );
 
     // Pattern to break string in separate words
-    final static Pattern wordPattern = Pattern.compile("[^\\W]+");
+    final static Pattern WORD_PATTERN = Pattern.compile("[^\\W]+");
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:105]: If `WORD_PATTERN` is used within this class only, then it's better to make it private.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java [hunk 3]
```diff
@@ -747,4 +774,30 @@ else if (actionConfiguration.getBody().matches("(?s).*,[\\W]*?(?i)WHERE.*")) {
         }
         return actionConfiguration;
     }
+
+    private Mono<PageDTO> sendGenerateCRUDPageAnalyticsEvent(PageDTO page, Datasource datasource, String pluginName) {
+        return Mono.zip(
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:779]: Instead of using `zip`, this can be simplified with 
`sessionUserService.getCurrentUser().flatMap`

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java [hunk 4]
```diff
@@ -747,4 +774,30 @@ else if (actionConfiguration.getBody().matches("(?s).*,[\\W]*?(?i)WHERE.*")) {
...
+            final Map<String, Object> data = Map.of(
+                "applicationId", pageDTO.getApplicationId(),
+                "pageId", pageDTO.getId(),
+                "pageName", pageDTO.getName(),
+                "pluginName", pluginName,
+                "datasourceId", datasource.getId(),
+                "organizationId", datasource.getOrganizationId()
+            );
+            analyticsService.sendEvent(AnalyticsEvents.GENERATE_CRUD_PAGE.getEventName(), currUser.getUsername(), data);
+            return Mono.just(page);
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:795]: This can be simplified with:
`return analyticsService.sendEvent(AnalyticsEvents.GENERATE_CRUD_PAGE.getEventName(), currUser.getUsername(), data).thenReturn(page);`

**Review by nayan-rafiq — CHANGES_REQUESTED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 1 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:90]: Assuming this may be required by other parts of the application, we can also move this to a separate constant file.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 2 above)
Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:105]: This pattern looks odd. If I read this right, we are using this pattern to match _every word_ in the given text, right? Then the pattern could be simplified to just `\w+`, no? Also, if my interpretation is right, can we change the comment above as well to read something like `Pattern to match all words in the text`. The current comment sounds like we are using this to _break_ a string up into parts, as opposed to match something.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java [hunk 5]
```diff
@@ -279,19 +289,30 @@
                     )
                     .collect(Collectors.toList());
 
-                log.debug("Going to update layout for page {0} and layout {1}", savedPageId.get(), layoutId);
+                // Extract S3 bucket name from template application and map to users bucket. Bucket name is stored at
+                // index 1 in plugin specified templates
+                if (S3_PLUGIN_PACKAGE.equals(plugin.getPackageName()) && !CollectionUtils.isEmpty(templateActionList)) {
+                    mappedColumnsAndTableName.put(
+                        templateActionList.get(0).getUnpublishedAction().getActionConfiguration().getPluginSpecifiedTemplates().get(1).getValue().toString(),
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:296]: Can we move the hard-coded `1` here to a variable similar to how it was done in the S3 plugin? So, if we change this in the S3 plugin in the future, we have some indication of it's use here as well. Currently, it'll be *very* hard to find this lone `1` here and identify what it actually means.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 4 above)
Comment (abhvsn) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:795]: sendEvent is not returning any reactive type like flux or mono so I guess this will not be possible. Am I missing something here?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 1 above)
Comment (abhvsn) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:90]: Hey thanks for the tip 👍

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java [hunk 6]
```diff
@@ -668,18 +693,20 @@ private JSONObject updateTemplateWidgets(JSONObject widgetDsl, Map<String, Strin
                 }
             } else if (FieldName.DROP_DOWN_WIDGET.equals(widgetDsl.getAsString(FieldName.TYPE))
                 && FieldName.OPTIONS.equals(key)
-                && !defaultDropdownValue.toLowerCase().contains("asc")) {
-                // This will handle the options field in SelectWidget
+                && !(defaultDropdownValue.toLowerCase().contains("asc") || defaultDropdownValue.equals("1"))) {
```
Comment (sharat87) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:696]: Is there a reason we are doing `.contains` here, instead of equality checking? Wouldn't `defaultDropdownValue.equalsIgnoreCase("asc")` work?

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 4 above)
Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:795]: I'm not sure that would work, since the `.sendEvent` method doesn't return a `Mono`. It's a `void` returning method. 🤔

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 6 above)
Comment (abhvsn) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:696]: Yes changing the implementation here

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 2 above)
Comment (abhvsn) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:105]: About the regex pattern, I want to match every word and then map to the required word using the Maps. [Here](https://regex101.com/r/qIF2co/1) is the tryout to match and replace words can you take a look and revert. Sorry, I'm not good at regex 😅.
Earlier I thought here we are just compiling the pattern and not using any matcher but the ultimate goal is to match every word from the string. I'll update the comment.

Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:105]: No worries man, I should've put in some more detail in my message. Let me try again.

The current regex says `[^\W]+`, in English, this reads `one or more of (anything that is not (not a word char))`. You see the two consecutive `not`s in there? They negate each other. The `\W` matches any non-word characters, like spaces, commas, dots etc. The `\w` matches any word characters, like alphabets, digits, etc.

The regex `\w+`, in English, reads `one or more of (a word char)`, and so it just matches all words. Here's a table that might shed some more light into how this thing works:

| Pattern | English |
| --- | --- |
| `[^\W]+` | one or more of (anything that is not (not a word char)) |
| `[\W]+` | one or more of (anything that is (not a word char)) |
| `[\w]+` | one or more of (anything that is (a word char)) |
| `\w+` | one or more of (a word char) |

It matches the same thing too:

Screenshot 2021-07-23 at 13 31 32

Let me know if I'm not making sense and we can get on a call 😃

Comment (abhvsn) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:105]: Thanks for the explanation 😄

**Review by sharat87 — DISMISSED** [review]
Looks good to be merged. Please squash merge once the CI is finished. Thanks!

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 4 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:797]: @abhvsn Sorry, my mistake. In that case, we shouldn't we use `map` instead of `flatMap` here @sharat87 ?

Comment (sharat87) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:797]: Yep, `.map` will work just fine.

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java [hunk 7]
```diff
@@ -747,4 +780,26 @@ else if (actionConfiguration.getBody().matches("(?s).*,[\\W]*?(?i)WHERE.*")) {
         }
         return actionConfiguration;
     }
+
+    private Mono<PageDTO> sendGenerateCRUDPageAnalyticsEvent(PageDTO page, Datasource datasource, String pluginName) {
```
Comment (nayan-rafiq) [inline app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:784]: This function may like the following:
private Mono<PageDTO> sendGenerateCRUDPageAnalyticsEvent(PageDTO page, Datasource datasource, String pluginName) {
        return sessionUserService.getCurrentUser()
        .map(currUser -> {

            final Map<String, Object> data = Map.of(
                "applicationId", page.getApplicationId(),
                "pageId", page.getId(),
                "pageName", page.getName(),
                "pluginName", pluginName,
                "datasourceId", datasource.getId(),
                "organizationId", datasource.getOrganizationId()
            );
            analyticsService.sendEvent(AnalyticsEvents.GENERATE_CRUD_PAGE.getEventName(), currUser.getUsername(), data);
            return page;
        })
        .onErrorResume(e -> {
            log.warn("Error sending generate CRUD DB table page data point", e);
            return page;
        });
    }

**Review by nayan-rafiq — CHANGES_REQUESTED** [review]

File: app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java (hunk 7 above)
Comment (nayan-rafiq) [reply app/server/appsmith-server/src/main/java/com/appsmith/server/solutions/CreateDBTablePageSolution.java:784]: Also please change the currUser -> currentUser

**Review by nayan-rafiq — APPROVED** [review]
